### ISSUE TITLE
Splink 3.0 Comparison Columns interface

### ISSUE THREAD

**Issue by RobinL** [moj-analytical-services/splink#262]
Currently, each item of the `comparison_columns` list in the settings object looks like this:

```
{
    "num_levels": 3,
    "term_frequency_adjustments": true,
    "col_name": "name",
    "gamma_index": 0,
    "data_type": "string",
    "fix_u_probabilities": false,
    "fix_m_probabilities": false,
    "case_expression": "case when name_l is null or name_r is null then -1 when jaro_winkler_sim(name_l, name_r) >= 1.0 then 2 when jaro_winkler_sim(name_l, name_r) >= 0.88 then 1 else 0 end as gamma_name",
    "m_probabilities": [ 0.1, 0.2, 0.7 ],
    "u_probabilities": [ 0.7, 0.2, 0.1 ],
    "tf_adjustment_weights": [ 0.0, 0.0, 1.0 ]
}
```

This suffers from a number of problems for users:

- If the user specifies num_levels, but no case expression, Splink automagically chooses a sql CASE expression
- If the user specifies a case expression, they must hand code a compatible `num_levels`
- The m an u probabilities implicitly align to the non-null levels in the case expression, so user must cross reference the two
- Same for tf adjustment weights
- Not possible to specify different columns for tf adjustments depending on level of case statement 
- data type is superfluous
- user has to specify col_name.  in case of single column, this is superfluous.
- custom columns must be used when user wants to use multiple columns in their case expression.  
- not possible to associate labels with individual levels of a comparison columns, meaning charts cannot be labelled 
- The SQL statement is hard to read when saved out as json because it's often a long complex multiline case statement within the `case_expression` key

In Splink 3.0, I propose to simply this.  My proposal is that all comparison columns will be a list, with each element specified as follows:

```
[
{
    "col_name": "my_name",
    "comparison_levels": [
        {
            "sql_expr": "WHEN name_l IS NULL OR name_r IS NULL THEN -1",
            "label": "null level"             
        },
        {
            "sql_expr": "WHEN JARO_WINKLER_SIM(name_l, name_r) >= 0.95 THEN 2",
            "label": "jaro > 0.95"
        },
        {
            "sql_expr": "WHEN JARO_WINKLER_SIM(name_l, name_r) >= 0.88 THEN 1",
            "label": "jaro > 0.88"
        },
        {
            "sql_expr": "ELSE 0",
            "label": "other"
        }
    ],
},
{
...other comparsion columns...
}

```

For each level within `comparison_levels`, the following will be optional:
- `label`.  If specified, charts will be labelled with user provided input.  Otherwise autopopulated with `level_0` etc.
- `tf_adjustment_column`. If specified, term frequencies applied to matches on this level using column specified
- `tf_adjustment_weight`.  If specified, and specified, term frequencies weighted using this value.  Otherwise assumed to be 1 where `tf_adjustment_column` specified.

A fully complete level (with much of this information completed automatically by Splink) would be something like:
```
 {
            "sql_expr": "WHEN JARO_WINKLER_SIM(name_l, name_r) >= 0.95 THEN 2",
            "label": "jaro > 0.95",
            "u_probability": 0.2,
            "m_probability": 0.1,
            "tf_adjustment_column": "name",
            "tf_adjustment_weight": 0.8
        },
```

The following information will _no longer_ need to be provided by the user:
- `num_levels`
- `col_name` will be optional
- `custom_columns_used` inferred from the data
- `data type` will be removed

*Important* 

The functions in `splink.case_statements` will be converted to auto-produce a 'comarison_levels' dict for the user.  So in simple cases, the user will populate the settings dictionary from these 'library' case statements rather than having to author the full dict themselves.

Questions:
- Is this an improvement?
- Should there be an option for the user to provide a `case_expression` which, under the hood, is translated into `comparison_levels` as above.  Alternatively, should there be a splink function that does this to provide a starting point for the user
- Should we retain the current shorthand for term frequency adjustments as a boolean.  If True, applies tf adjustments to the highest level?

The proposal will have significant benefits in terms of simplifying the codebase, preventing user error, and making outputs clearer

**Comment by RobinL** [comment]
Some advantages purely from user experience point of view:
- With the new interface, the user won't be able to have a setting dictionary that contradicts itself (e.g. num levels, or m and u probabilities with wrong number of values)
- Easier to give user warnings and other information about what they've done wrong e.g. if they haven't included a null level
- Fewer surprises.  I think there's too much magic at the moment - particuarly autoselecting case expression for the user based on 'num levels'

From a functionality point of view this will mean:
- User able to do more sophisticated term frequency adjustments with different columns dependning on level within a single comparison column
- Easier to label charts
- Will be able to delete quite a lot of code.
-

**Comment by samnlindsay** [comment]
I think I broadly agree with all of the above.

One observation is that `"sql_expr"` could be simplified further, in keeping with the ambition of this change and DRY principles: 
:x: `WHEN JARO_WINKLER_SIM(name_l, name_r) >= 0.95 THEN 2`
✅  `JARO_WINKLER_SIM(name_l, name_r) >= 0.95`

If the point is for the user to describe each level with a label and a condition, rather than to explicitly be building a complete CASE WHEN... expression, then the latter is what I would expect to be provided - _"where did `2` come from?🤔 "_

If your current proposal assumes the `comparison_levels` list must be ordered and levels numbered as shown (i.e. level -1 has to be listed first, and `other` has to be listed last) then I can foresee introducing new possibilities for user error:
- Copy and paste - two expressions with the same `THEN n`
- Out of order:
  - Null level not listed first
  - `ELSE 0` not listed last (**Sidenote:** is it not a requirement that there be a `ELSE n END` to ensure no comparisons go uncounted? In that case, could it not be inserted automatically?)

Instead, could the user provide:
- a list of *conditions* for each level
- an optional *label* (default `level_0` etc.) 
- an optional *order* of evaluation (default, order provided)
- an optional `null_level` flag, labelled as such by default

And could "level 0" automatically be populated using `ELSE 0 END` and labelled `other`, which to me seems unambiguous and harmless?

### Suggestion
```
{
    "col_name": "my_name",
    "comparison_levels": [
         {
            "sql_condition": "JARO_WINKLER_SIM(name_l, name_r) >= 0.95",
            "label": "jaro > 0.95",
            "order": 1
        },
        {
            "sql_condition": "JARO_WINKLER_SIM(name_l, name_r) >= 0.88",
            "label": "jaro > 0.88",
            "order": 2
        },        
        {
            "sql_expr": "name_l IS NULL OR name_r IS NULL",
            "null_level": true      
        },
    ],
},
{
...other comparsion columns...
}
```

- Order of list doesn't matter, and you can arbitrarily add new levels without worrying about inconsistencies
- Can provide informative warnings/error message:
  - "You have not specified a null level - are you sure?"
  - "{label_1} and {label_2} columns have the same value for 'order' - this must be unique for each level"

**Comment by RobinL** [comment]
I like it.  It would mean that, in addition to having a library of full definitions of comparison columns e.g. `from splink.comparison_columns import jaro_3` the library would also define comparison levels.

so for instance, `jaro_3` may itself be defined as:

```
def exact_match(col_name):
   return {'sql_condition': f"{col_name}_r = {col_name}_l", "label":exact_match"}

...

def jaro_3(col_name):
    return [exact_match(col_name), jaro_level(col_name, 0.95), else_level()]


```


Only part of your proposal I'm not sure about is `order`, for two reasons:
-  I'm not 100% sure that the null condition should always come first in the case statement. 
-  I think allowing the user to specify levels out of order, with an (optional) order key is potentially confusing)

Instead I wonder if it's simpler to require that `comparison_levels`  appears in the order of evaluation.

**Comment by mamonu** [comment]
Some thoughts  (sorry if they are a bit disjointed)

I like where this is heading.
In a way this also means that no radical change will be needed on the scala jar which believe it or not makes me happier (at this point in time).

On one hand I would be interested to sort of take a lot of the logic from the case expressions into the jar itself. 

But... to dynamically be able to do that is quite an undertaking (ie provide parameters in pyspark. execute in scala) that I would not like to think about while there are deadlines to follow etc. And its going to be an overhaul thats as deep as... splink v1 to v2 or v3 so nothing that should be taken lightly. Not impossible. But nothing that can be whipped up quickly. 

Also I think that one advantage of splink IMO in is the power the custom case expressions are giving to linkers via simple config json "commands".

I like the proposed splink v3 changes to streamline the custom case API and would prefer this being the focus rather than getting more logic into the jar. This in a way obfuscates a lot of the inner-workings.

This is also something that will be transferable if at any point splink is also going to have another backend other than spark.
Not everything needing linking is millions of rows big.Adding more logic to the jar will make porting to a different backend more difficult.

**Comment by samnlindsay** [comment]
> I like it. It would mean that, in addition to having a library of full definitions of comparison columns e.g. `from splink.comparison_columns import jaro_3` the library would also define comparison levels.
> 
> so for instance, `jaro_3` may itself be defined as:
> 
> ```
> def exact_match(col_name):
>    return {'sql_condition': f"{col_name}_r = {col_name}_l", "label":exact_match"}
> 
> ...
> 
> def jaro_3(col_name):
>     return [exact_match(col_name), jaro_level(col_name, 0.95), else_level()]
> ```
> 
Love this, I hadn't thought of that but it's a very neat alternative for people who don't want to have to go down the custom SQL route. It also means as we develop new custom case expressions for our own work, we can easily generalise them and add to the splink library.

> Instead I wonder if it's simpler to require that `comparison_levels` appears in the order of evaluation.
Not sure what you mean here...

**Comment by RobinL** [comment]
@mamonu Agreed.  Another aspect to this is that the discovery of `sqlglot` means we should be able to auto-translate between:  spark, duckdb and presto (Athena),  sqlite, Postgres, Mysql and others.

This autotranslation will work best if we try and keep our SQL fairly contrained and simple - which I think argues against putting more logic into the jar.

This would enable e.g. us to take a model trained with Spark, and output a low-latency sqlite version of the same model (that can ingest a pairwise comparison and output a score).  

It will also mean that for small datasets, people can use DuckDB as a backend rather than Spark, which should substantially improve the appeal of Splink to a greater range of users.

**Comment by RobinL** [comment]
@samnlindsay   

To clarify the "require that comparison_levels appears in the order of evaluation."

All I mean is that the user provides an array (list) of comparison columns.

What I'm suggesting is that the order of items in this list matters, and we enforce this rather than allowing the user to specify e.g. `  "order": 2`.

So the first element in the comparison columns list is always the first `when` statement in the `case when` SQL statement.

The only real complexity here is the null level, where, as you suggested, the user would have to set `  "null_level": true     ` if they didn't want m and u weights to be calculated for the column.  We can alert the user with a warning if they haven't set `null_level`  in any comparison columns.

**Comment by RobinL** [comment]
Closed by #278
